runtime/runtime/src/global_contracts.rs

```diff
@@ -0,0 +1,232 @@
+            GlobalContractIdentifier::AccountId(account_id.clone())
+        }
+    };
+    let distribution_receipts = Receipt::new_global_contract_distribution(
+        account_id,
+        GlobalContractDistributionReceipt::new(
+            id,
+            // we start with any shard, using the current one just for convenience
+            current_shard_id,
+            vec![],
+            contract_code,
+        ),
+    );
+    result.new_receipts.push(distribution_receipts);
+}
+
+fn apply_distribution_current_shard(
+    receipt: &Receipt,
+    global_contract_data: &GlobalContractDistributionReceipt,
+    apply_state: &ApplyState,
+    state_update: &mut TrieUpdate,
+) {
+    let config = apply_state.config.wasm_config.clone();
+    let trie_key = TrieKey::GlobalContractCode {
+        identifier: match &global_contract_data.id() {
+            GlobalContractIdentifier::CodeHash(hash) => {
+                GlobalContractCodeIdentifier::CodeHash(*hash)
+            }
+            GlobalContractIdentifier::AccountId(account_id) => {
+                GlobalContractCodeIdentifier::AccountId(account_id.clone())
+            }
+        },
+    };
+    state_update.set(trie_key, global_contract_data.code().to_vec());
+    state_update.commit(StateChangeCause::ReceiptProcessing { receipt_hash: receipt.get_hash() });
+    let code_hash = match global_contract_data.id() {
+        GlobalContractIdentifier::CodeHash(hash) => Some(*hash),
+        GlobalContractIdentifier::AccountId(_) => None,
+    };
+    let _ = precompile_contract(
+        &ContractCode::new(global_contract_data.code().to_vec(), code_hash),
+        config,
+        apply_state.cache.as_deref(),
+    );
+}
+
+fn forward_distribution_next_shard(
+    receipt: &Receipt,
+    global_contract_data: &GlobalContractDistributionReceipt,
+    apply_state: &ApplyState,
+    epoch_info_provider: &dyn EpochInfoProvider,
+    state_update: &mut TrieUpdate,
+    receipt_sink: &mut ReceiptSink,
+) -> Result<(), RuntimeError> {
+    let shard_layout = epoch_info_provider.shard_layout(&apply_state.epoch_id)?;
+    let already_delivered_shards = BTreeSet::from_iter(
+        global_contract_data
+            .already_delivered_shards()
+            .iter()
+            .cloned()
+            .chain(std::iter::once(apply_state.shard_id)),
+    );
+    if let Some(next_shard) = shard_layout
+        .shard_ids()
+        .filter(|shard_id| !already_delivered_shards.contains(&shard_id))
+        .next()
+    {
+        let mut next_receipt = Receipt::new_global_contract_distribution(
+            receipt.predecessor_id().clone(),
+            GlobalContractDistributionReceipt::new(
+                global_contract_data.id().clone(),
+                next_shard,
+                Vec::from_iter(already_delivered_shards),
+                global_contract_data.code().clone(),
+            ),
+        );
+        let receipt_id = apply_state.create_receipt_id(receipt.receipt_id(), 0);
```

Comment:
Is it fine to use 0 for receipt index? Can you remind me what is receipt index?

Reply:
receipt index is a sequence number in the list of all child receipts of a given receipt.
for action receipts it looks like this:
```
        let receipt_ids = result
            .new_receipts
            .into_iter()
            .enumerate()
            .filter_map(|(receipt_index, mut new_receipt)| {
                let receipt_id = create_receipt_id_from_receipt_id(
                    apply_state.current_protocol_version,
                    receipt.receipt_id(),
                    ...
                    receipt_index,
                );
```